x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/IndicesAccessControl.java

```diff
@@ -21,11 +21,53 @@
  */
 public class IndicesAccessControl {
 
-    public static final IndicesAccessControl ALLOW_ALL = new IndicesAccessControl(true, Collections.emptyMap());
-    public static final IndicesAccessControl ALLOW_NO_INDICES = new IndicesAccessControl(true,
-            Collections.singletonMap(IndicesAndAliasesResolverField.NO_INDEX_PLACEHOLDER,
-                    new IndicesAccessControl.IndexAccessControl(true, new FieldPermissions(), DocumentPermissions.allowAll())));
-    public static final IndicesAccessControl DENIED = new IndicesAccessControl(false, Collections.emptyMap());
+    public static final IndicesAccessControl ALLOW_ALL = new IndicesAccessControl(true, null) {
+        @Override
+        public IndexAccessControl getIndexPermissions(String index) {
+            assert false == Regex.isSimpleMatchPattern(index) :
+                    "index access control can and should only be retrieved for a given concrete name, but requested [" + index + "]";
+            return IndexAccessControl.PROMISCUOUS_ACCESS_CONTROL;
+        }
+    };
+    public static final IndicesAccessControl ALLOW_NO_INDICES = new IndicesAccessControl(true, null) {
+        @Override
+        public IndexAccessControl getIndexPermissions(String index) {
+            for (String placeHolderIndexName : IndicesAndAliasesResolverField.NO_INDICES_OR_ALIASES_ARRAY) {
+                if (placeHolderIndexName.equals(index)) {
+                    return IndexAccessControl.PROMISCUOUS_ACCESS_CONTROL;
+                }
+            }
+            assert false : "the \"no index access control\" is empty and no access control can be retrieved, but requested [" + index + "]";
+            return null;
+        }
+    };
+    public static final IndicesAccessControl DENIED = new IndicesAccessControl(false, null) {
+
+        IndexAccessControl DENIED_ACCESS_CONTROL = new IndexAccessControl(false, null, null) {
```

Comment:
Because this is a special type of `IndexAccessControl` (it does not support most of its operations) and it is used only in this particular situation, I don't think we should make it `public`. If we need thee same index access control behaviour with another occasion we can make the move then.

Reply:
I was *not* suggesting to make it `public`. My suggestion was to move it inside `IndexAccessControl`. So the structure would be:

```java
public class IndicesAccessControl {
     
    public static final IndicesAccessControl DENIED = new IndicesAccessControl(false, ...) { ... };  

    public static class IndexAccessControl {
        IndexAccessControl DENIED_ACCESS_CONTROL = new IndexAccessControl(false, ...) { ... };
    }
}
```

instead of

```java
public class IndicesAccessControl {
     
    public static final IndicesAccessControl DENIED = new IndicesAccessControl(false, ...) { 
        IndexAccessControl DENIED_ACCESS_CONTROL = new IndexAccessControl(false, ...) { ... };
        ...
    };  

    public static class IndexAccessControl {
        ...
    }
}
```

Reply:
I have moved `DENIED_ACCESS_CONTROL` inside `IndexAccessControl`.